### PR TITLE
Detect circular instantiated types

### LINKED ISSUE #3309: Stack overflow for infinitely recursive anonymous type
```ts
function foo<T>() {
    var z = foo<typeof y>();
    var y: {
        y2: typeof z
    };
    return y;
}


function bar<T>() {
    var z = bar<typeof y>();
    var y: {
        y2: typeof z;
    }
    return y;
}

var a = foo<number>();
var b = bar<number>();
a = b;
```
Because the type parameter T is in scope when we encounter the anonymous type for `y`, a new type is created for the type of `y`. But each time foo/bar is called, yet a different type for `y` is created inside that call!

@ahejlsberg and I are leaning towards a solution that classifies such anonymous types (with type parameters in scope) as References, and give them targets. That way, isDeeplyNestedGeneric will bottom out and detect recursion.

We may need a similar mechanism in the display writer.

### PR DESCRIPTION
**Pull request by ahejlsberg**
Fixes #3309 (and revises fix for #3237).

This PR switches our `isDeeplyNestedGeneric` function to track both type references and instantiated anonymous types. The key change is that we use the symbol associated with the type as the tracked identity (which is always the symbol of the type being instantiated). We likewise now use symbols to track cyclic anonymous types in the type-to-string logic.

The PR removes the anonymous type instantiation caching implemented in #3239 as it is no longer necessary.

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -4313,12 +4306,12 @@ module ts {
             // Effectively, we will generate a false positive when two types are structurally equal to at least 10 levels, but unequal at
             // some level beyond that.
             function isDeeplyNestedGeneric(type: ObjectType, stack: ObjectType[]): boolean {
-                if (type.flags & TypeFlags.Reference && depth >= 10) {
-                    let target = (<TypeReference>type).target;
+                if (type.flags & (TypeFlags.Reference | TypeFlags.Instantiated) && depth >= 10) {
+                    let symbol = type.symbol;
```
Comment (JsonFreeman) [inline src/compiler/checker.ts]: I don't think this is necessary because you only refer to `symbol` once

Comment (ahejlsberg) [inline src/compiler/checker.ts]: Not sure what you mean. It is referenced in a loop that will run at least 10 times (as we known depth >= 10).

File: src/compiler/types.ts [hunk 2]
```diff
@@ -1582,14 +1582,15 @@ module ts {
         Tuple                   = 0x00002000,  // Tuple
         Union                   = 0x00004000,  // Union
         Anonymous               = 0x00008000,  // Anonymous
+        Instantiated            = 0x00010000,  // Instantiated anonymous type
```
Comment (JsonFreeman) [inline src/compiler/types.ts]: Can you rename this InstantiatedAnonymous? Or have all the types created by instantiateType take on the flag as well.

Also, it seems to me that References are always Instantiated. To that end, can you OR this flag into the Reference flag as well so the subset / superset relation is clear?

File: src/compiler/checker.ts (hunk 1 above)
Comment (JsonFreeman) [inline src/compiler/checker.ts]: I mean just once lexically. I didn't know the runtime references count individually wrt to local vars.

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -4313,12 +4306,12 @@ module ts {
...
             function isDeeplyNestedGeneric(type: ObjectType, stack: ObjectType[]): boolean {
-                if (type.flags & TypeFlags.Reference && depth >= 10) {
-                    let target = (<TypeReference>type).target;
+                if (type.flags & (TypeFlags.Reference | TypeFlags.Instantiated) && depth >= 10) {
+                    let symbol = type.symbol;
                     let count = 0;
                     for (let i = 0; i < depth; i++) {
                         let t = stack[i];
-                        if (t.flags & TypeFlags.Reference && (<TypeReference>t).target === target) {
+                        if (t.flags & (TypeFlags.Reference | TypeFlags.Instantiated) && t.symbol === symbol) {
```
Comment (JsonFreeman) [inline src/compiler/checker.ts]: Please put a comment here saying what Reference and Instantiated each refer to.

**Comment by JsonFreeman** [comment]
:

File: src/compiler/types.ts (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/compiler/types.ts]: Agreed.  'Instantiated' conveys something entirely different to me.  Furthermore, if it is supposed to mean InstantiatedAnonymousTYpe, then i'm surprised that it doesn't include 0x00008000 as part of it.

**Pull request merged**